server/aesdsocket.c

```diff
@@ -0,0 +1,594 @@
+#define FILE_PERMISSIONS (0644)
+#define BUFF_SIZE        100
+#define FILE_PATH        "/var/tmp/aesdsocketdata"
+
+typedef struct
+{
+	int clifd;
+	pthread_mutex_t *mutex;
+	pthread_t thread_id;
+	bool thread_complete;
+}thread_parameter;
+
+//Linked list node
+struct slist_data_s
+{
+	thread_parameter	thread_params;
+	SLIST_ENTRY(slist_data_s) entries;
+};
+typedef struct slist_data_s slist_data_t;
+pthread_mutex_t mutex_lock = PTHREAD_MUTEX_INITIALIZER;
+
+
+/*
+ *Global variables
+ */
+int sockfd, clifd, filefd;
+int status;
+char *cli_buff;
+int packet_length = RESET;
+char rx_buff[BUFF_SIZE];
+slist_data_t 	*data_node 		= NULL;
+/*
+ *Function Name: signal_handler (int signal)
+ *Description: handles signal occurered during execution
+ *return: void
+ */
+static void signal_handler(int signal)
+{
+	switch (signal)	// signal handler
+	{
+		case SIGINT:
+			syslog(LOG_INFO, "SIGINT signal received\n");
+			printf("SIGINT occured\n");
+			break;
+
+		case SIGTERM:
+			syslog(LOG_INFO, "SIGTERM signal received\n");
+			printf("SIGTERM occured\n");
+			break;
+
+		case SIGKILL:
+			syslog(LOG_INFO, "SIGkill signal received\n");
+			printf("SIGKILL occured\n");
+			break;
+
+		default:
+			syslog(LOG_ERR, "Unknown Signal received\n");
+			exit(EXIT_FAILURE);
+			break;
+
+	}
+
+	// clear and free buffers
+	syslog(LOG_INFO, "Clearing buffers and Exiting\n");
+	printf("Clearing buffers and Exiting\n");
+//	free(cli_buff);
+	unlink(FILE_PATH);
+	close(sockfd);
+	close(clifd);
+	exit(EXIT_SUCCESS);
+}
+
+/*
+ *Function Name: void* thread_handler(void* thread_param)
+ *Description: Handles thread specific functionalities
+ *return: int
+ */
+void* thread_handler(void* thread_param){
+	int i;
+	bool packet_status = false;
+	ssize_t rx_data = RESET;
+	ssize_t wr_data = RESET;
+	char char_read = RESET;
+	
+	thread_parameter * t_params = (thread_parameter *) thread_param;
+	  //Allocate appropriate buffer memory
+		cli_buff = (char*) malloc((sizeof(char) *BUFF_SIZE));
+		if (cli_buff == NULL)
+		{
+			syslog(LOG_ERR, "Failed to allocate memory\n");
+			printf("malloc() failed\n");
+			exit(EXIT_FAILURE);
+		}
+		else
+		{
+			syslog(LOG_INFO, "Memory allocation successfull\n");
+			printf("malloc() succed\n");
+		}
+
+		memset(cli_buff, RESET, BUFF_SIZE);
+		
+		while (!packet_status)
+		{
+			// Recieve data
+			rx_data = recv(clifd, rx_buff, BUFF_SIZE, 0);
+			if (rx_data < 0)
+			{
+				syslog(LOG_ERR, "Failed to receive:%d\n", status);
+				printf("Receive data failed\n");
+				exit(EXIT_FAILURE);
+			}
+			else if (rx_data == 0)
+			{
+				syslog(LOG_INFO, "Reception successfull\n");
+				printf("Received data successfully\n");
+			}
+
+			//Traverse through end of packet
+			for (i = 0; i < BUFF_SIZE; i++)
+			{
+				if (rx_buff[i] == '\n')
+				{
+					packet_status = true;
+					i++;
+					break;
+				}
+			}
+
+			packet_length += i;
+
+			//reallocate memory to appropriate size 
+			cli_buff = (char*) realloc(cli_buff, (packet_length + 1));
```

Comment:
@varun2506 what assumption is made here about the size of RAM heap relative to the total size of all packets sent?